In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
import pandas as pd
accident = pd.read_csv('datasets\\accident_data.csv')

In [3]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
accident.info

<bound method DataFrame.info of                 Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1  

In [5]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [6]:
accident.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [7]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [8]:
accident['Latitude']=accident['Latitude'].fillna(accident['Latitude'].mode()[0])
accident['Longitude']=accident['Longitude'].fillna(accident['Longitude'].mode()[0])
accident['Road_Surface_Conditions']=accident['Road_Surface_Conditions'].fillna(accident['Road_Surface_Conditions'].mode()[0])
accident['Road_Type']=accident['Road_Type'].fillna(accident['Road_Type'].mode()[0])
accident['Urban_or_Rural_Area']=accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])

In [9]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions        0
Road_Type                      0
Urban_or_Rural_Area            0
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [10]:
accident['Accident Data']=pd.to_datetime(accident['Accident Date'],dayfirst= True,errors='coerce')

In [11]:
accident.dtypes

Index                              object
Accident_Severity                  object
Accident Date                      object
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
Accident Data              datetime64[ns]
dtype: object

In [12]:
accident['Weather_Conditions']=accident['Weather_Conditions'].fillna('unacccounted')
accident.isnull().sum()

Index                           0
Accident_Severity               0
Accident Date                   0
Latitude                        0
Light_Conditions                0
District Area                   0
Longitude                       0
Number_of_Casualties            0
Number_of_Vehicles              0
Road_Surface_Conditions         0
Road_Type                       0
Urban_or_Rural_Area             0
Weather_Conditions              0
Vehicle_Type                    0
Accident Data              395672
dtype: int64

<h1>What is the distribution of accident severity levels?</h1>

In [13]:
severity_distribution = accident['Accident_Severity'].value_counts()
severity_distribution


Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h1>Insight 1 Accidents are predominantly of a 'Slight' nature: A significant majority, approximately 86.5% of the accidents (563,801 out of 651,679 total accidents), fall under the 'Slight' category, indicating that most incidents result in minor injuries or damages.</h1>

<h1>Inisght 1.1 Severe accidents (Serious and Fatal combined) make up a smaller proportion: Together, 'Serious' and 'Fatal' accidents account for roughly 13.5% of all cases, with 'Serious' accidents being about 13.5% and 'Fatal' incidents only 1.3%. This highlights that while severe cases are less common, they still represent a notable concern for safety improvements.</h1>

<h1>Q2 How frequently do accidents occur in urban versus rural areas?</h1>

In [14]:
urban_rural_distribution = accident['Urban_or_Rural_Area'].value_counts()
urban_rural_distribution


Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h1>Insight 2 Accidents are significantly more frequent in urban areas. Approximately 64% of the reported accidents (421,678 out of 660,679 total) occur in urban settings compared to 36% in rural areas. This suggests that the higher population density, increased traffic, and more complex road networks in urban regions contribute to the elevated number of incidents.</h1>

<h1>Q3   Average Number of Casualties for Day Light Conditions</h1>

In [15]:
weather_distribution = accident['Weather_Conditions'].value_counts()
weather_distribution




Weather_Conditions
Fine no high winds       520885
Raining no high winds     79696
Other                     17150
unacccounted              14128
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h1>Insight 3 Clear weather accounts for the highest number of casualties: "Fine no high winds" conditions overwhelmingly dominate the figures, with 520,885 casualties. This suggests that good weather, while generally safer, also coincides with higher traffic volumes and potentially riskier behaviors, contributing to the majority of incidents.</h1>

<h1>Q4   What is the average number of casualties per accident?</h1>

In [16]:
average_casualties = accident['Number_of_Casualties'].mean()
average_casualties


np.float64(1.357040257068864)

<h1>Insight 4   On average, there are approximately 1.36 casualties per accident. This indicates that most accidents tend to involve a single individual, but a notable number of incidents likely involve multiple casualties, highlighting the potential severity of many situations.</h1>

<h1>Q5 How frequently do accidents occur under different light conditions?</h1>

In [17]:
light_conditions_distribution = accident['Light_Conditions'].value_counts()
light_conditions_distribution


Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h1>Insight 5   This will show how often accidents occur in daylight or darkness.</h1>

<h1>Q6  How does light condition affect the number of casualties in accidents?</h1>

In [18]:
light_casualties = accident.groupby('Light_Conditions')['Number_of_Casualties'].mean()
light_casualties 

Light_Conditions
Darkness - lighting unknown    1.306447
Darkness - lights lit          1.369374
Darkness - lights unlit        1.331105
Darkness - no lighting         1.541657
Daylight                       1.340309
Name: Number_of_Casualties, dtype: float64

<h1>Insight 6  Poor lighting conditions increase the number of casualties per accident. For instance, accidents in "Darkness - no lighting" have the highest average casualties at 1.54, indicating that the absence of lighting greatly exacerbates the severity of accidents.</h1>

<h1>Q7  How do road surface conditions affect the number of vehicles involved in accidents?</h1>

In [19]:
surface_vehicles = accident.groupby('Road_Surface_Conditions')['Number_of_Vehicles'].mean()
surface_vehicles

Road_Surface_Conditions
Dry                     1.841631
Flood over 3cm. deep    1.610619
Frost or ice            1.640006
Snow                    1.746859
Wet or damp             1.829161
Name: Number_of_Vehicles, dtype: float64

<h1>Insight7  Dry and wet/damp conditions have the highest average number of vehicles involved in accidents, which suggests that while adverse conditions like snow, ice, and floods cause cautious driving, they still present significant risks.</h1>

<h1>Q8  How frequently do accidents occur on different road types</h1>

In [20]:
road_type_distribution = accident['Road_Type'].value_counts()
road_type_distribution



Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

<h1>Insight 8.1  Single carriageways are by far the most accident-prone road type with 496,663 incidents recorded. This significantly surpasses all other road types combined, indicating that single carriageways may present unique safety challenges, such as narrower lanes, mixed traffic, or higher speeds in some areas.</h1>

<h1>Insight 8.2 Slip roads have the fewest accidents at just 704 recorded incidents. This suggests they might be relatively safer due to their specific design, which often includes controlled access and slower speeds compared to other road types.</h1>

<h1>Q9 What is the most common vehicle type involved in accidents?</h1>

In [32]:
vehicle_type_distribution = accident['Vehicle_Type'].value_counts()
vehicle_type_distribution


Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h1>Insight 9.1 Cars dominate accident involvement, with 497,992 cases, accounting for the vast majority. This isn't surprising given cars are likely the most common mode of transportation, but it also highlights their significant contribution to road incidents.</h1>

<h1>Insight 9.2 Vans and light goods vehicles are the second-most frequent vehicle type involved, with 34,160 incidents, far below cars but still notable. This may be due to their regular use in delivery and logistics, increasing their exposure on the roads</h1>

<h1>Insight 9.3 Motorcycles show varying levels of involvement based on engine size, with larger motorcycles (over 500cc) involved in 25,657 accidents, while smaller motorcycles (50cc and under) only account for 7,603 incidents. This trend may be related to differences in speed, usage patterns, or rider demographics.</h1>

<h1>Q10  How many accidents occur in each district area?</h1>

In [22]:
district_area_distribution = accident['District Area'].value_counts()
district_area_distribution


District Area
Birmingham            13491
Leeds                  8898
Manchester             6720
Bradford               6212
Sheffield              5710
                      ...  
Berwick-upon-Tweed      153
Teesdale                142
Shetland Islands        133
Orkney Islands          117
Clackmannanshire         91
Name: count, Length: 422, dtype: int64

<h1>Insight 10  This will show the number of accidents in different district areas.</h1>

<h1>Q11 What is the correlation between road surface conditions and the number of vehicles involved?</h1>

In [23]:
urban_casualties = accident.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].mean()
urban_casualties

Urban_or_Rural_Area
Rural          1.479204
Unallocated    1.181818
Urban          1.287807
Name: Number_of_Casualties, dtype: float64

<h1>Insight 11  Rural road surface conditions show the highest correlation (1.479204) with the number of vehicles involved in accidents, compared to urban areas (1.287807) and unallocated conditions (1.181818). This suggests that rural environments may pose unique challenges, such as higher speeds, less lighting, or varying maintenance levels, which could contribute to a greater likelihood of accidents involving multiple vehicles.</h1>

<h1>Q12 How frequently do accidents occur on different days of the week?</h1>

In [24]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst=True, errors='coerce')
accident['Day_of_Week'] = accident['Accident Date'].dt.day_name()
day_of_week_distribution = accident['Day_of_Week'].value_counts()
day_of_week_distribution


Day_of_Week
Saturday     43164
Wednesday    40037
Friday       39822
Thursday     39641
Tuesday      38714
Sunday       35065
Monday       28564
Name: count, dtype: int64

<h1>Insight12.1 Saturday has the highest number of accidents (43,164), followed closely by Wednesday (40,037). This could indicate that Saturdays, being part of the weekend, may involve more leisure travel or higher vehicle volume, potentially contributing to a higher accident rate.</h1>

<h1>Insight12.2 Monday sees the fewest accidents (28,564), significantly lower than other days of the week. This might be attributed to more structured weekday routines and potentially less non-essential travel compared to weekends.</h1>

<h1>Q13 How are accident dates distributed across the months?</h1>

In [25]:
accident['Month'] = accident['Accident Date'].dt.month
month_distribution = accident['Month'].value_counts()
month_distribution


Month
11.0    24240
12.0    24156
10.0    23962
7.0     22939
9.0     22558
2.0     22264
6.0     22196
3.0     21824
5.0     21723
8.0     21106
4.0     19787
1.0     18252
Name: count, dtype: int64

<h1>Insight13 This will reveal if certain months have higher accident rates, possibly due to weather or seasonal variations.</h1>

<h1>How do road surface conditions vary between urban and rural areas?</h1>

In [26]:
urban_surface_conditions = pd.crosstab(accident['Urban_or_Rural_Area'], accident['Road_Surface_Conditions'])
urban_surface_conditions


Road_Surface_Conditions     Dry  Flood over 3cm. deep  Frost or ice  Snow  \
Urban_or_Rural_Area                                                         
Rural                    145145                   785         10953  3102   
Unallocated                   5                     1             0     0   
Urban                    303397                   231          7564  2788   

Road_Surface_Conditions  Wet or damp  
Urban_or_Rural_Area                   
Rural                          79005  
Unallocated                        5  
Urban                         107698

<h1>This will show if certain surface conditions are more prevalent in urban or rural settings.</h1>

<h1>Q15 What is the average number of vehicles involved in accidents for each district area?</h1>

In [27]:
district_vehicles = accident.groupby('District Area')['Number_of_Vehicles'].mean()
district_vehicles



District Area
Aberdeen City    1.637188
Aberdeenshire    1.553368
Adur             1.969305
Allerdale        1.782801
Alnwick          1.646552
                   ...   
Wychavon         1.828068
Wycombe          1.866397
Wyre             1.860258
Wyre Forest      1.789474
York             1.856089
Name: Number_of_Vehicles, Length: 422, dtype: float64

<h1>Insight 15 This will reveal if certain district areas have a higher average number of vehicles involved in accidents.</h1>

<h1>Q16 What is the relationship between road type, weather conditions, and the number of casualties?</h1>

In [30]:
q16=accident.groupby(['Road_Type', 'Weather_Conditions'])['Number_of_Casualties'].mean()
q16




Road_Type           Weather_Conditions   
Dual carriageway    Fine + high winds        1.442396
                    Fine no high winds       1.477963
                    Fog or mist              1.665689
                    Other                    1.433746
                    Raining + high winds     1.475652
                    Raining no high winds    1.502453
                    Snowing + high winds     1.389189
                    Snowing no high winds    1.399020
                    unacccounted             1.330494
One way street      Fine + high winds        1.183544
                    Fine no high winds       1.185132
                    Fog or mist              1.241379
                    Other                    1.233010
                    Raining + high winds     1.307190
                    Raining no high winds    1.242127
                    Snowing + high winds     1.071429
                    Snowing no high winds    1.230769
                    unacccounted        

<h1>Insight 16.1 Weather conditions significantly influence accident risks across all road types, with certain combinations amplifying dangers. For example, fog or mist consistently correlates with higher casualty rates on various road types, particularly on dual carriageways (1.665689). This suggests that adverse weather conditions, especially those impairing visibility, play a critical role in increasing the likelihood of accidents with casualties, regardless of the road type.</h1>

<h1>Insight 16.2 Complex road designs coupled with severe weather conditions create challenging driving environments. For instance, roundabouts show high correlations with casualties during snowing with high winds (1.533333). This highlights how weather factors combined with intricate navigation requirements can exacerbate accident risks, emphasizing the need for heightened caution and perhaps better infrastructure planning for such conditions.</h1>